### PR TITLE
bpf: Don't perform L3 operation when ENABLE_ROUTING is disabled

### PR DESCRIPTION
**Pull request by tgraf**
The datapath was incorrectly performing an L3 operation even if routing was
disabled for packets falling through and being passed to the stack.

Reported-by: @rabysh
Signed-off-by: Thomas Graf <thomas@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9549)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

File: bpf/bpf_lxc.c [hunk 1]
```diff
@@ -69,7 +69,9 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 				   struct ipv6_ct_tuple *tuple, int l3_off,
 				   struct ipv6hdr *ip6, __u32 *dstID)
 {
+#ifdef ENABLE_ROUTING
 	union macaddr router_mac = NODE_MAC;
+#endif
```
Comment (borkmann) [inline bpf/bpf_lxc.c:74]: Small nit, feel free to ignore: we could potentially reduce scope as in https://github.com/cilium/cilium/blob/master/bpf/bpf_lxc.c#L115 which also consolidates into single ifdef.

**Review by borkmann — APPROVED** [review]
Lacking a bit of context what was reported by @rabysh (would be great to have some more info in commit msg for future reference), but overall change looks okay.

**Comment by jraby** [comment]
Context:

The following  happens when deploying cilium in `generic-veth` cni chaining mode, along with the  `bridge` cni plugin as the first in the chain.
example cni config:
```
{
    "cniVersion": "0.3.0",
    "name": "generic-veth",
    "plugins": [
        {
            "bridge": "cni0",
            "hairpinMode": true,
            "ipam": {
                "subnet": "10.205.3.0/24",
                "type": "host-local"
            },
            "isDefaultGateway": true,
            "name": "kubernetes",
            "type": "bridge"
        },
        {
            "type": "cilium-cni"
        }
    ]
}
```

With that config, packets being sent from a pod will have their destination MAC rewritten to something that is invalid.

For example:

```
19:12:00.046501 5a:58:fa:47:7d:11 > 3a:e4:e5:b1:31:9f, ethertype IPv4 (0x0800), length 98: (tos 0x0, ttl 64, id 23472, offset 0, flags [DF], proto ICMP (1), length 84)
    10.205.3.1 > 10.205.3.29: ICMP echo request, id 13332, seq 540, length 64

19:12:00.046551 3a:e4:e5:b1:31:9f > 42:1e:d5:69:fe:70, ethertype IPv4 (0x0800), length 98: (tos 0x0, ttl 63, id 50299, offset 0, flags [none], proto ICMP (1), length 84)
    10.205.3.29 > 10.205.3.1: ICMP echo reply, id 13332, seq 540, length 64

^^^ this packets never reaches  10.205.3.1 since the dst MAC is wrong
```

In the above:
- `cni0` on the host has ip `10.205.3.1` and mac `5a:58:fa:47:7d:11`
- `eth0` in the pod has ip `10.205.3.29` and mac `3a:e4:e5:b1:31:9f`
- the `vethxxxxx` interface that links cni0 with eth0 in the pod has mac `42:1e:d5:69:fe:70` (and no ip)

The trace shows packets going from cni0 to eth0 with the proper mac addresses, but the return packet has `42:1e:d5:69:fe:70` as its dst mac, which is wrong. It should be `5a:58:fa:47:7d:11`. (and thus, the packet get lost / it never reaches `cni0`)

Also, in the pod, the arp table and `ip neigh` information show the right mac address for 10.205.3.1:
```$ sudo nsenter -n -t 48657   ip neigh
10.205.3.1 dev eth0 lladdr 5a:58:fa:47:7d:11 STALE

$ sudo nsenter -n -t 48657   ip route
default via 10.205.3.1 dev eth0
10.205.3.0/24 dev eth0 proto kernel scope link src 10.205.3.29
```


With the change, the mac address is not rewritten anymore, so packets flow properly.

**Pull request merged**